cpp/src/dbscan/runner.cuh

```diff
@@ -65,57 +67,80 @@ void final_relabel(Index_* db_cluster, Index_ N, cudaStream_t stream,
     [MAX_LABEL] __device__(Index_ val) { return val == MAX_LABEL; }, allocator);
 }
 
-/* @param N number of points
- * @param D dimensionality of the points
- * @param eps epsilon neighborhood criterion
- * @param minPts core points criterion
- * @param labels the output labels (should be of size N)
- * @param ....
- * @param temp temporary global memory buffer used to store intermediate computations
- *             If this is a null pointer, then this function will return the workspace size needed.
- *             It is the responsibility of the user to cudaMalloc and cudaFree this buffer!
- * @param stream the cudaStream where to launch the kernels
- * @return in case the temp buffer is null, this returns the size needed.
+/**
+ * Run the DBSCAN algorithm (common code for single-GPU and multi-GPU)
+ * @tparam opg Whether we are running in a multi-node multi-GPU context
+ * @param[in]  handle       raft handle
+ * @param[in]  x            Input data (N*D row-major device array)
+ * @param[in]  N            Number of points
+ * @param[in]  D            Dimensionality of the points
+ * @param[in]  start_row    Index of the offset for this node
+ * @param[in]  n_owned_rows Number of rows (points) owned by this node
```

Comment:
If we can assume this is always equal to `N` for the single-GPU case, this might be more of a reason to abstract the SG & MG implementations behind an interface and maybe provide this as a constructor arg for the MG impl. That would also give more reason to having separate public API functions for MG & SG (since the argument lists will differ).

Reply:
This is not part of the public API, but rather a helper function for `dbscanFitImpl`. So I am not really bothered by the variables `N` and `n_owned_rows` being redundant in the SG case, as it allows us to have a single code path for SG and MG (at the exception of two communication blocks that are called only in MG case and that I have now separated in their own files).

I see the point of refactoring the DBSCAN code because the current design is far from ideal, but it sounds like a larger refactoring that doesn't belong in this PR. or maybe I misunderstood what you want exactly.

Reply:
I was mostly building upon my comment [here](https://github.com/rapidsai/cuml/pull/3382#discussion_r564137906). Just thinking this could be cleaner if we used classes here, something like a `dbscan_impl_t` abstract class that might accept a struct `dbscan_param_t` in its constructor and implement all the common functionality, then an additional implementation `opg_dbscan_impl_t` which could potentially extend `dbscan_impl_t` but accept the extra args in the constructor and compose its own run method by calling functions from the super class but lacing the opg-specific logic where it belongs. This is just a quick example off the top of my head, but my thinking is that it would separate the concerns between the SG and OPG implementations while still keeping one consistent logical flow from the APIs which invoke it.

I agree, this could be a lot to do in this PR, especially so close to burndown if we are planning to get this into 0.18. Mind creating a quick issue for it just so it doesn't get lost? It doesn't have to be the design I mentioned, but it would clean up the runner a bit if the OPG logic was in separate functions at least.

Reply:
#3422